File: app/src/main/java/com/benny/pxerstudio/pxerexportable/ExportingUtils.java
package com.benny.pxerstudio.pxerexportable;

import android.content.Context;
import android.media.MediaScannerConnection;
import android.net.Uri;
import android.os.Environment;
import android.support.annotation.NonNull;
import android.support.constraint.ConstraintLayout;
import android.view.LayoutInflater;
import android.widget.EditText;
import android.widget.SeekBar;
import android.widget.TextView;

import com.afollestad.materialdialogs.DialogAction;
import com.afollestad.materialdialogs.GravityEnum;
import com.afollestad.materialdialogs.MaterialDialog;
import com.benny.pxerstudio.R;
import com.benny.pxerstudio.util.Tool;
import com.benny.pxerstudio.widget.PxerView;

import java.io.File;
import java.util.List;

/**
 * Created by BennyKok on 10/17/2016.
 */
public class ExportingUtils {
    private static ExportingUtils ourInstance = new ExportingUtils();

    public static ExportingUtils getInstance() {
        return ourInstance;
    }

    public MaterialDialog currentProgressDialog;

    private ExportingUtils() {
    }

    public void dismissAllDialogs(){
        if(currentProgressDialog != null)
            currentProgressDialog.dismiss();
    }

    public File checkAndCreateProjectDirs(){
        String path = Environment.getExternalStorageDirectory().getPath().concat("/PxerStudio/Export");
        File dirs = new File(path);
        if (!dirs.exists()) {
            dirs.mkdirs();
        }
        return dirs;
    }

    public static String getExportPath(){
        return Environment.getExternalStorageDirectory().getPath().concat("/PxerStudio/Export");
    }

    public static String getProjectPath(){
        return Environment.getExternalStorageDirectory().getPath().concat("/PxerStudio/Project");
    }

    public File checkAndCreateProjectDirs(String extraFolder){
        if (extraFolder == null || extraFolder.isEmpty())return checkAndCreateProjectDirs();
        String path = Environment.getExternalStorageDirectory().getPath().concat("/PxerStudio/Export/"+extraFolder);
        File dirs = new File(path);
        if (!dirs.exists()) {
            dirs.mkdirs();
        }
        return dirs;
    }

    public void toastAndFinishExport(Context context,String fileName){
        if (fileName != null && !fileName.isEmpty())
        MediaScannerConnection.scanFile(context,
                new String[]{fileName}, null,
                new MediaScannerConnection.OnScanCompletedListener() {
                    public void onScanCompleted(String path, Uri uri) {}
                });

        Tool.toast(context,"Exported successfully");
    }

    public void scanAlotsOfFile(Context context,List<File> files){
        String[] paths = new String[files.size()];
        for (int i = 0; i < files.size(); i++) {
            paths[i] = files.get(i).toString();
        }
        MediaScannerConnection.scanFile(context,
                paths, null,
                new MediaScannerConnection.OnScanCompletedListener() {
                    public void onScanCompleted(String path, Uri uri) {}
                });
    }

    public void showProgressDialog(Context context){
        currentProgressDialog = new MaterialDialog.Builder(context)
                .titleGravity(GravityEnum.CENTER)
                .typeface(Tool.myType, Tool.myType)
                .cancelable(false)
                .canceledOnTouchOutside(false)
                .title("Painting...")
                .progress(true, 0)
                .progressIndeterminateStyle(true)
                .show();
    }

    public void showExportingDialog(final Context context, final PxerView pxerView, final OnExportConfirmedListenser listenser){
        showExportingDialog(context,-1,pxerView,listenser);
    }

    public void showExportingDialog(final Context context,int maxSize, final PxerView pxerView, final OnExportConfirmedListenser listenser){
        final ConstraintLayout l = (ConstraintLayout) LayoutInflater.from(context).inflate(R.layout.dialog_activity_drawing, null);
        final EditText editText = (EditText) l.findViewById(R.id.et1);
        final SeekBar seekBar = (SeekBar) l.findViewById(R.id.sb);
        final TextView textView = (TextView) l.findViewById(R.id.tv2);

        editText.setText(pxerView.getProjectName());

        if (maxSize == -1)
            seekBar.setMax(4096 - pxerView.getPicWidth());
        else
            seekBar.setMax(maxSize - pxerView.getPicWidth());

        textView.setText("Size : " + String.valueOf(pxerView.getPicWidth()) + " x " + String.valueOf(pxerView.getPicHeight()));
        seekBar.setOnSeekBarChangeListener(new SeekBar.OnSeekBarChangeListener() {
            @Override
            public void onProgressChanged(SeekBar seekBar, int i, boolean b) {
                textView.setText("Size : " + String.valueOf(i + pxerView.getPicWidth()) + " x " + String.valueOf(i + pxerView.getPicHeight()));
            }

            @Override
            public void onStartTrackingTouch(SeekBar seekBar) {
            }

            @Override
            public void onStopTrackingTouch(SeekBar seekBar) {
            }
        });

        new MaterialDialog.Builder(context)
                .titleGravity(GravityEnum.CENTER)
                .typeface(Tool.myType, Tool.myType)
                .customView(l, false)
                .title("Export")
                .positiveText("Export")
                .negativeText("Cancel")
                .onPositive(new MaterialDialog.SingleButtonCallback() {
                    @Override
                    public void onClick(@NonNull MaterialDialog dialog, @NonNull DialogAction which) {
                        if (editText.getText().toString().isEmpty()){
                            Tool.toast(context,"The file name cannot be empty!");
                            return;
                        }
                        listenser.OnExportConfirmed(editText.getText().toString(),seekBar.getProgress() + pxerView.getPicWidth(),seekBar.getProgress() + pxerView.getPicHeight());
                    }
                })
                .show();
    }

    public interface OnExportConfirmedListenser{
       public void OnExportConfirmed(String fileName,int width,int height);
    }
}

File: app/src/main/java/com/benny/pxerstudio/pxerexportable/ExportingUtils.kt
package com.benny.pxerstudio.pxerexportable

import android.content.Context
import android.media.MediaScannerConnection
import android.view.LayoutInflater
import android.view.View
import android.widget.EditText
import android.widget.SeekBar
import android.widget.SeekBar.OnSeekBarChangeListener
import android.widget.TextView
import androidx.constraintlayout.widget.ConstraintLayout
import com.afollestad.materialdialogs.MaterialDialog
import com.afollestad.materialdialogs.customview.customView
import com.benny.pxerstudio.R
import com.benny.pxerstudio.util.Tool.toast
import com.benny.pxerstudio.widget.PxerView
import java.io.File

/**
 * Created by BennyKok on 10/17/2016.
 */
object ExportingUtils {
    @JvmField
    var currentProgressDialog: MaterialDialog? = null
    fun dismissAllDialogs() {
        if (currentProgressDialog != null) currentProgressDialog!!.dismiss()
    }

    fun checkAndCreateProjectDirs(context: Context): File {
        val path = context.getExternalFilesDir("/")!!.path + "/PxerStudio/Export"
        val dirs = File(path)
        if (!dirs.exists()) {
            dirs.mkdirs()
        }
        return dirs
    }

    fun checkAndCreateProjectDirs(extraFolder: String?, context: Context): File {
        if (extraFolder == null || extraFolder.isEmpty()) return checkAndCreateProjectDirs(context)
        val path = context.getExternalFilesDir("/")!!.path + "/PxerStudio/Export/" + extraFolder
        val dirs = File(path)
        if (!dirs.exists()) {
            dirs.mkdirs()
        }
        return dirs
    }

    fun toastAndFinishExport(context: Context?, fileName: String?) {
        if (fileName != null && fileName.isNotEmpty()) MediaScannerConnection.scanFile(
            context, arrayOf(fileName), null
        ) { _, _ -> }
        toast(context, "Exported successfully")
    }

    fun scanAlotsOfFile(context: Context?, files: List<File>) {
        val paths = arrayOfNulls<String>(files.size)
        for (i in files.indices) {
            paths[i] = "${files[i]}"
        }
        MediaScannerConnection.scanFile(
            context,
            paths, null
        ) { _, _ -> }
    }

    fun showProgressDialog(context: Context?) {
        currentProgressDialog = MaterialDialog(context!!)
//            .typeface(Tool.myType, Tool.myType)
            .cancelable(false)
            .cancelOnTouchOutside(false)
            .title(null, "Painting...")
            .message(null, "Exporting...")
//            .progress
//            .progress(true, 0)
//            .progressIndeterminateStyle(true)

        currentProgressDialog!!.show()
    }

    fun showExportingDialog(
        context: Context?,
        pxerView: PxerView,
        listener: OnExportConfirmedListener
    ) {
        showExportingDialog(context, -1, pxerView, listener)
    }

    fun showExportingDialog(
        context: Context?,
        maxSize: Int,
        pxerView: PxerView,
        listener: OnExportConfirmedListener
    ) {
        val l = LayoutInflater.from(context)
            .inflate(R.layout.dialog_activity_drawing, null) as ConstraintLayout
        val nameEdit = l.findViewById<View>(R.id.dialog_drawing_name_edit) as EditText
        val sizeSeekBar = l.findViewById<View>(R.id.dialog_drawing_size_seekBar) as SeekBar
        val sizeTv = l.findViewById<View>(R.id.dialog_drawing_size) as TextView
        nameEdit.setText(pxerView.projectName)
        if (maxSize == -1) sizeSeekBar.max = 4096 - pxerView.picWidth else sizeSeekBar.max =
            maxSize - pxerView.picWidth
        sizeTv.text =
            "Size : " + java.lang.String.valueOf(pxerView.picWidth) +
                    " x " + java.lang.String.valueOf(
                pxerView.picHeight
            )
        sizeSeekBar.setOnSeekBarChangeListener(object : OnSeekBarChangeListener {
            override fun onProgressChanged(seekBar: SeekBar, i: Int, b: Boolean) {
                sizeTv.text =
                    "Size : " + (i + pxerView.picWidth).toString() +
                            " x " + (i + pxerView.picHeight).toString()
            }

            override fun onStartTrackingTouch(seekBar: SeekBar) {}
            override fun onStopTrackingTouch(seekBar: SeekBar) {}
        })
        MaterialDialog(context!!)
//            .titleGravity(GravityEnum.CENTER)
//            .typeface(Tool.myType, Tool.myType)
            .customView(view = l)
            .title(null, "Export")
            .positiveButton(null, "Export") {
                if (nameEdit.text.toString().isEmpty()) {
                    toast(context, "The file name cannot be empty!")
                    return@positiveButton
                }
                listener.onExportConfirmed(
                    nameEdit.text.toString(),
                    sizeSeekBar.progress + pxerView.picWidth,
                    sizeSeekBar.progress + pxerView.picHeight
                )
            }
            .negativeButton(null, "Cancel")
            .show()
    }

    interface OnExportConfirmedListener {
        fun onExportConfirmed(fileName: String?, width: Int, height: Int)
    }

    fun getExportPath(context: Context): String {
        return context.getExternalFilesDir("/")!!.path + "/PxerStudio/Export"
    }

    fun getProjectPath(context: Context): String {
        return context.getExternalFilesDir("/")!!.path + "/PxerStudio/Project"
    }
}
